pkg/controller/pv_controller.go

```diff
@@ -0,0 +1,125 @@
+/*
+ Copyright 2024 Juicedata Inc
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/juicedata/juicefs-csi-driver/pkg/config"
+	"github.com/juicedata/juicefs-csi-driver/pkg/k8sclient"
+)
+
+var (
+	pvCtrlLog = klog.NewKlogr().WithName("pv-controller")
+
+	// used secret set
+	watchedSecrets = map[string]struct{}{}
+)
+
+type PVController struct {
+	*k8sclient.K8sClient
+}
+
+func NewPVController(client *k8sclient.K8sClient) *PVController {
+	return &PVController{client}
+}
+
+func (m *PVController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	pvCtrlLog.V(1).Info("Receive pv", "name", request.Name)
+	pv, err := m.GetPersistentVolume(ctx, request.Name)
+	if err != nil {
+		pvCtrlLog.Error(err, "Failed to get pv", "name", request.Name)
+		return reconcile.Result{}, err
+	}
+	if pv.Spec.CSI != nil && pv.Spec.CSI.NodePublishSecretRef != nil {
+		secretName := pv.Spec.CSI.NodePublishSecretRef.Name
+		secretNamespace := pv.Spec.CSI.NodePublishSecretRef.Namespace
+		if _, ok := watchedSecrets[secretName]; !ok {
+			watchedSecrets[fmt.Sprintf("%s/%s", secretNamespace, secretName)] = struct{}{}
+			sc := NewSecretController(m.K8sClient)
+			_, err := sc.Reconcile(ctx, reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Namespace: secretNamespace,
+					Name:      secretName,
+				},
+			})
```

Comment:
make a function called `RefreshSecret`, and called by both here and secretController.